File: sdk/akg-ts/test/conformance.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { Store } from '../src/store.js';
import { NotFoundError, InvalidInputError, MissingRequiredFieldError } from '../src/errors.js';

const CONFORMANCE_DIR = join(import.meta.dirname ?? __dirname, '../../../testdata/conformance');
const MANIFEST_PATH = join(CONFORMANCE_DIR, 'manifest.json');

interface StoreExpectation {
  nodes: number;
  edges: number;
  has_uncompacted_wal: boolean;
  next_wal_sequence: number;
  absent_node?: { type: string; id: string };
}

interface Fixture {
  path: string;
  purpose: string;
  expected_result: 'accept' | 'reject';
  validation_scope: 'format' | 'store';
  store_expectation?: StoreExpectation;
  expected_error_category?: string;
  sha256?: string;
  features?: string[];
}

interface Manifest {
  version: number;
  fixtures: Fixture[];
}

function errorCategory(err: unknown): string {
  if (!(err instanceof Error)) return 'unknown';
  const msg = err.message.toLowerCase();

  if (msg.includes('invalid header') && (msg.includes('magic') || msg.includes('too short') || msg.includes('version') || msg.includes('reserved') || msg.includes('checksum algorithm'))) {
    return 'invalid_header';
  }
  if (msg.includes('wal checksum mismatch')) return 'wal_checksum_mismatch';
  if (msg.includes('checksum mismatch')) return 'checksum_mismatch';
  if (msg.includes('invalid section table') || msg.includes('wrong section counts') || msg.includes('section too short')) return 'invalid_section_table';
  if (msg.includes('invalid section ranges') || msg.includes('sections overlap') || msg.includes('out of bounds')) return 'invalid_section_ranges';
  if (msg.includes('invalid bloom section')) return 'invalid_bloom_section';
  if (msg.includes('unknown wal operation')) return 'unknown_wal_operation';
  if (msg.includes('wal checksum mismatch')) return 'wal_checksum_mismatch';
  if (msg.includes('invalid wal record') || msg.includes('non-increasing sequence')) return 'invalid_wal_record';
  if (msg.includes('invalid wal payload')) return 'invalid_wal_payload';
  if (msg.includes('invalid data payload')) return 'invalid_data_payload';
  if (msg.includes('derived index mismatch')) return 'derived_index_mismatch';
  if (msg.includes('malformed') && msg.includes('key')) return 'malformed_key';
  return 'unknown';
}

const manifest: Manifest = JSON.parse(readFileSync(MANIFEST_PATH, 'utf-8'));

describe('conformance', () => {
  for (const fixture of manifest.fixtures) {
    it(fixture.path + ' — ' + fixture.purpose, () => {
      const filePath = join(CONFORMANCE_DIR, fixture.path);
      const buf = readFileSync(filePath);
      const bytes = new Uint8Array(buf.buffer, buf.byteOffset, buf.byteLength);

      if (fixture.expected_result === 'accept') {
        let store: ReturnType<typeof Store.fromBytes>;
        expect(() => {
          store = Store.fromBytes(bytes, filePath);
        }, `fixture ${fixture.path} should open without error`).not.toThrow();

        if (fixture.store_expectation && fixture.validation_scope === 'store') {
          const exp = fixture.store_expectation;
          const nodes = store!.listNodes();
          const edges: unknown[] = [];
          for (const node of nodes) {
            const out = store!.outboundEdges({ type: node.type, id: node.id });
            edges.push(...out);
          }

          expect(nodes.length, `node count for ${fixture.path}`).toBe(exp.nodes);

          const edgeCount = countEdges(store!);
          expect(edgeCount, `edge count for ${fixture.path}`).toBe(exp.edges);

          expect(store!.hasUncompactedWAL, `has_uncompacted_wal for ${fixture.path}`).toBe(exp.has_uncompacted_wal);
          expect(Number(store!.nextWALSequence), `next_wal_sequence for ${fixture.path}`).toBe(exp.next_wal_sequence);

          if (exp.absent_node) {
            const n = store!.getNode(exp.absent_node.type, exp.absent_node.id);
            expect(n, `absent node in ${fixture.path}`).toBeNull();
          }
        }
      } else {
        let caughtError: unknown;
        try {
          Store.fromBytes(bytes, filePath);
        } catch (e) {
          caughtError = e;
        }
        expect(caughtError, `fixture ${fixture.path} should throw`).toBeDefined();

        if (fixture.expected_error_category) {
          const category = errorCategory(caughtError);
          expect(category, `error category for ${fixture.path}: ${caughtError instanceof Error ? caughtError.message : String(caughtError)}`).toBe(fixture.expected_error_category);
        }
      }
    });
  }
});

function countEdges(store: ReturnType<typeof Store.fromBytes>): number {
  const nodes = store.listNodes();
  const seen = new Set<string>();
  for (const node of nodes) {
    const edges = store.outboundEdges({ type: node.type, id: node.id });
    for (const e of edges) {
      const key = `${e.from.type}:${e.from.id}:${e.relation}:${e.to.type}:${e.to.id}`;
      seen.add(key);
    }
  }
  return seen.size;
}

File: sdk/akg-go/conformance_test.go
package akg

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"testing"
)

type conformanceManifest struct {
	Fixtures []conformanceFixture `json:"fixtures"`
}

type conformanceFixture struct {
	Path                  string            `json:"path"`
	ExpectedResult        string            `json:"expected_result"`
	ExpectedErrorCategory string            `json:"expected_error_category"`
	ValidationScope       string            `json:"validation_scope"`
	StoreExpectation      *storeExpectation `json:"store_expectation"`
}

type storeExpectation struct {
	Nodes             int         `json:"nodes"`
	Edges             int         `json:"edges"`
	HasUncompactedWAL bool        `json:"has_uncompacted_wal"`
	NextWALSequence   uint64      `json:"next_wal_sequence"`
	AbsentNode        *absentNode `json:"absent_node"`
}

type absentNode struct {
	Type string `json:"type"`
	ID   string `json:"id"`
}

func TestConformance(t *testing.T) {
	fixtureDir := filepath.Join("..", "..", "testdata", "conformance")
	manifestData, err := os.ReadFile(filepath.Join(fixtureDir, "manifest.json"))
	if err != nil {
		t.Fatalf("read manifest: %v", err)
	}
	var manifest conformanceManifest
	if err := json.Unmarshal(manifestData, &manifest); err != nil {
		t.Fatalf("parse manifest: %v", err)
	}

	for _, fx := range manifest.Fixtures {
		fx := fx
		t.Run(fx.Path, func(t *testing.T) {
			data, err := os.ReadFile(filepath.Join(fixtureDir, fx.Path))
			if err != nil {
				t.Fatalf("read fixture: %v", err)
			}

			var openErr error
			var st *Store
			if fx.ValidationScope == "format" {
				_, openErr = decodeContainer(data)
			} else {
				st, openErr = openBytes(data)
			}

			switch fx.ExpectedResult {
			case "accept":
				if openErr != nil {
					t.Fatalf("expected accept, got error: %v", openErr)
				}
				if fx.StoreExpectation != nil && st != nil {
					checkStoreExpectation(t, st, fx.StoreExpectation)
				}
			case "reject":
				if openErr == nil {
					t.Fatal("expected reject, got nil error")
				}
				got := errorCategory(openErr)
				if got != fx.ExpectedErrorCategory {
					t.Fatalf("expected error category %q, got %q (error: %v)", fx.ExpectedErrorCategory, got, openErr)
				}
			default:
				t.Fatalf("unknown expected_result: %q", fx.ExpectedResult)
			}
		})
	}
}

func checkStoreExpectation(t *testing.T, st *Store, exp *storeExpectation) {
	t.Helper()
	if got := len(st.state.nodes); got != exp.Nodes {
		t.Errorf("node count: want %d, got %d", exp.Nodes, got)
	}
	if got := len(st.state.edges); got != exp.Edges {
		t.Errorf("edge count: want %d, got %d", exp.Edges, got)
	}
	if got := len(st.committedWAL) > 0; got != exp.HasUncompactedWAL {
		t.Errorf("has_uncompacted_wal: want %v, got %v", exp.HasUncompactedWAL, got)
	}
	if got := uint64(st.nextWALSeq); got != exp.NextWALSequence {
		t.Errorf("next_wal_sequence: want %d, got %d", exp.NextWALSequence, got)
	}
	if exp.AbsentNode != nil {
		node, err := st.GetNode(exp.AbsentNode.Type, exp.AbsentNode.ID)
		if err != nil {
			t.Errorf("GetNode for absent check: %v", err)
		} else if node != nil {
			t.Errorf("expected node {type=%q id=%q} to be absent", exp.AbsentNode.Type, exp.AbsentNode.ID)
		}
	}
}

func errorCategory(err error) string {
	switch {
	case errors.Is(err, errInvalidHeader):
		return "invalid_header"
	case errors.Is(err, errChecksumMismatch):
		return "checksum_mismatch"
	case errors.Is(err, errInvalidSectionRanges):
		return "invalid_section_ranges"
	case errors.Is(err, errInvalidBloomSection):
		return "invalid_bloom_section"
	case errors.Is(err, errInvalidSectionTable):
		return "invalid_section_table"
	case errors.Is(err, errInvalidWALPayload):
		return "invalid_wal_payload"
	case errors.Is(err, errWALChecksumMismatch):
		return "wal_checksum_mismatch"
	case errors.Is(err, errUnknownWALOperation):
		return "unknown_wal_operation"
	case errors.Is(err, errInvalidWALRecord):
		return "invalid_wal_record"
	case errors.Is(err, errDerivedIndexMismatch):
		return "derived_index_mismatch"
	case errors.Is(err, errInvalidDataPayload):
		return "invalid_data_payload"
	case errors.Is(err, errMalformedKey):
		return "malformed_key"
	default:
		return "unknown"
	}
}
